### PR TITLE
feat: add captcha security for admin login

### LINKED ISSUE #231: feat: Thêm CAPTCHA bảo mật trang đăng nhập admin (Cloudflare Turnstile / Google reCAPTCHA v3)
## Mô tả

Trang đăng nhập admin (`/admin/login`) hiện chỉ có rate limiting (8 req/min) nhưng chưa có cơ chế phân biệt bot với người thật. Cần thêm CAPTCHA để tăng khả năng chống brute-force và credential stuffing.

## Vấn đề

- **Brute-force:** Attacker có thể thử nhiều mật khẩu trong giới hạn rate limit mà không bị chặn thêm.
- **Credential stuffing:** Dùng danh sách email/password bị lộ từ nguồn khác để tự động thử đăng nhập.
- **Bot automation:** Script tự động có thể submit form login liên tục từ nhiều IP khác nhau để bypass rate limit.

## Giải pháp

Tích hợp CAPTCHA vào `POST /auth/login`, hỗ trợ 2 provider:

### Cloudflare Turnstile (ưu tiên)
- Widget hiển thị trong form, thường tự động pass cho người dùng thật
- Privacy-friendly, không tracking
- Tích hợp tốt với hạ tầng Cloudflare hiện có (R2)

### Google reCAPTCHA v3 (fallback)
- Invisible, chạy ngầm khi submit
- Trả về score 0.0–1.0, từ chối nếu dưới ngưỡng `CAPTCHA_MIN_SCORE`

## Thiết kế kỹ thuật

**Backend:**
- `CaptchaService` gọi API xác minh của Cloudflare/Google trước khi xử lý login
- Error code mới: `CAPTCHA_FAILED` (HTTP 422)
- Điều khiển bằng `CAPTCHA_ENABLED` env var (mặc định `false` — backward compatible)

**Frontend:**
- `TurnstileWidget` component render widget Cloudflare, nút Submit disabled cho đến khi xác nhận
- `useRecaptchaV3` hook load script Google reCAPTCHA và execute khi submit
- Widget tự reset sau login thất bại

**Biến môi trường mới:**

| Biến | Mô tả |
|------|-------|
| `CAPTCHA_ENABLED` | Bật/tắt (`true`/`false`, mặc định `false`) |
| `CAPTCHA_PROVIDER` | `cloudflare` hoặc `google` |
| `CAPTCHA_SECRET_KEY` | Secret key từ dashboard provider |
| `CAPTCHA_MIN_SCORE` | Ngưỡng score tối thiểu (Google v3, mặc định `0.5`) |
| `VITE_CAPTCHA_PROVIDER` | Provider cho frontend (phải khớp backend) |
| `VITE_CAPTCHA_SITE_KEY` | Site key từ dashboard provider |

### Cloudflare Turnstile
1. Vào Cloudflare Dashboard → Turnstile → tạo site mới
2. Copy **Site Key** và **Secret Key**
3. Thêm vào `.env`:
```env
CAPTCHA_ENABLED=true
CAPTCHA_PROVIDER=cloudflare
CAPTCHA_SECRET_KEY=<secret-key>
VITE_CAPTCHA_PROVIDER=cloudflare
VITE_CAPTCHA_SITE_KEY=<site-key>
```

### Google reCAPTCHA v3
1. Vào Google reCAPTCHA Admin → tạo site v3
2. Copy **Site Key** và **Secret Key**
3. Thêm vào `.env`:
```env
CAPTCHA_ENABLED=true
CAPTCHA_PROVIDER=google
CAPTCHA_SECRET_KEY=<secret-key>
CAPTCHA_MIN_SCORE=0.5
VITE_CAPTCHA_PROVIDER=google
VITE_CAPTCHA_SITE_KEY=<site-key>
```

## Acceptance Criteria

- [ ] Khi `CAPTCHA_ENABLED=false` (mặc định): login hoạt động bình thường, không widget
- [ ] Khi Cloudflare Turnstile bật: widget hiển thị trong form; Submit disabled đến khi widget xác nhận; widget reset sau login sai
- [ ] Khi Google reCAPTCHA v3 bật: token được gửi kèm request, invisible với người dùng
- [ ] `POST /auth/login` không có token khi CAPTCHA bật → 422 `CAPTCHA_FAILED`
- [ ] `POST /auth/login` với token giả → 422 `CAPTCHA_FAILED`
- [ ] CI xanh (lint, typecheck, build, E2E)

## PR liên quan

Closes #230

### PR DESCRIPTION
**Pull request by nguyenhoangtin1404**
Closes #231

## Summary

- **Backend**: `CaptchaService` gọi Cloudflare Turnstile hoặc Google reCAPTCHA v3 API để xác minh token trước khi xử lý login; điều khiển bằng `CAPTCHA_ENABLED` (mặc định `false` — backward compatible)
- **Backend**: Thêm error code `CAPTCHA_FAILED` (HTTP 422); `LoginDto` nhận thêm `captcha_token?` optional
- **Frontend**: `CaptchaWidget` render Cloudflare Turnstile trong form đăng nhập; nút Submit bị disable cho đến khi widget xác nhận; widget tự reset sau login thất bại
- **Frontend**: `useRecaptchaV3` hook cho Google reCAPTCHA v3 (invisible, execute khi submit)
- **Docs**: `ENV.md` và `.env.example` cập nhật đầy đủ biến CAPTCHA mới

### Cloudflare Turnstile (khuyến nghị)
1. Vào [Cloudflare Turnstile Dashboard](https://dash.cloudflare.com/?to=/:account/turnstile) → tạo site mới
2. Lấy **Site Key** (frontend) và **Secret Key** (backend)
3. Thêm vào `.env`:
```env
CAPTCHA_ENABLED=true
CAPTCHA_PROVIDER=cloudflare
CAPTCHA_SECRET_KEY=<secret-key>

VITE_CAPTCHA_PROVIDER=cloudflare
VITE_CAPTCHA_SITE_KEY=<site-key>
```

### Google reCAPTCHA v3
1. Vào [Google reCAPTCHA Admin](https://www.google.com/recaptcha/admin) → tạo site v3
2. Lấy **Site Key** và **Secret Key**
3. Thêm vào `.env`:
```env
CAPTCHA_ENABLED=true
CAPTCHA_PROVIDER=google
CAPTCHA_SECRET_KEY=<secret-key>
CAPTCHA_MIN_SCORE=0.5

VITE_CAPTCHA_PROVIDER=google
VITE_CAPTCHA_SITE_KEY=<site-key>
```

## Test plan

- [ ] Khi `CAPTCHA_ENABLED=false` (mặc định): login hoạt động bình thường không có widget
- [ ] Khi Cloudflare Turnstile bật: widget hiển thị trong form; submit bị disable cho đến khi Turnstile xác nhận; widget reset khi login sai mật khẩu
- [ ] Khi Google reCAPTCHA v3 bật: login submit bình thường (invisible); token được gửi kèm request
- [ ] `POST /auth/login` không có token khi `CAPTCHA_ENABLED=true` → 422 `CAPTCHA_FAILED`
- [ ] `POST /auth/login` với token giả → 422 `CAPTCHA_FAILED`

https://claude.ai/code/session_018g98ZKWTjfZgWVuiq6fKxv

### CODE REVIEW

**Comment by nguyenhoangtin1404** [comment]
## Multi-role Deep Review (Tech Lead + PO + PM + Senior QA)\n\n### 1) Tech Lead review\n- **Strengths**\n  - Tách CaptchaModule/CaptchaService rõ ràng, điểm chặn đặt trước uthService.login là đúng kiến trúc.\n  - Hỗ trợ cả Turnstile + reCAPTCHA v3, có timeout và xử lý lỗi endpoint ngoài.\n  - FE UX tốt: disable submit khi chưa có token Turnstile, reset widget khi login fail.\n- **Major concerns**\n  - **Major**: frontend bật CAPTCHA theo VITE_CAPTCHA_SITE_KEY (CAPTCHA_ENABLED = !!siteKey), backend bật theo CAPTCHA_ENABLED === 'true'. Hai điều kiện bật/tắt tách rời dễ gây lệch cấu hình FE/BE trong production.\n    - **Khuyến nghị**: chuẩn hóa 1 source of truth (ví dụ backend /config/public hoặc quy ước cứng + startup validation fail-fast).\n  - **Major**: PR chưa cập nhật docs/API_CONTRACT.md cho payload login mới (captcha_token?) và lỗi CAPTCHA_FAILED (422).\n    - Theo invariant của repo, thay đổi contract phải sync docs cùng PR.\n  - **Major**: thiếu test backend cho CaptchaService (provider mismatch, secret thiếu, timeout, non-200, action mismatch, score threshold).\n\n### 2) PO review\n- Tính năng đúng intent bảo mật login admin và có backward compatibility (CAPTCHA_ENABLED=false).\n- Tuy nhiên acceptance chưa trọn vì thiếu phần contract docs + test chứng minh hành vi ở edge cases.\n\n### 3) PM review\n- Scope tốt nhưng chưa merge-ready cho môi trường production do risk cấu hình và coverage test còn thiếu.\n- Đề xuất tiêu chí Done trước merge:\n  1) Đồng bộ cơ chế bật/tắt CAPTCHA giữa FE/BE.\n  2) Cập nhật docs/API_CONTRACT.md.\n  3) Thêm unit tests cho CaptchaService + ít nhất 1 integration test login khi CAPTCHA bật.\n\n### 4) Senior QA (10+ years) review\n- **Test gaps quan trọng**\n  - Missing token khi CAPTCHA_ENABLED=true -> 422 (đã mô tả test plan, cần automation).\n  - Token invalid / provider outage / timeout / action != login / score dưới ngưỡng.\n  - Regression: khi CAPTCHA tắt, login flow phải giữ nguyên behavior cũ (status code, message, cookie path).\n\n### Verdict\n- **Request changes trước merge**: đây là PR bảo mật đăng nhập, nên cần khóa chặt docs contract + test coverage + đồng bộ config FE/BE trước khi approve.\n

File: backend/src/auth/auth.controller.ts [hunk 1]
```diff
@@ -94,17 +96,17 @@ export class AuthController {
   @UseInterceptors(LoginAuditInterceptor)
   async login(
     @Body() loginDto: LoginDto,
-    @Request() req,
+    @Request() req: ExpressRequest,
     @Res({ passthrough: true }) res: Response,
   ) {
+    await this.captchaService.verify(loginDto.captcha_token, req.ip);
```
[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply backend/src/auth/auth.controller.ts]: Addressed in latest commit: remote IP for CAPTCHA now follows the same audited request context in the login flow, and captcha gating stays before credential validation.

File: backend/src/common/captcha/captcha.service.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      this.logger.error('Google reCAPTCHA siteverify unreachable', err instanceof Error ? err.message : String(err));
+      throw new AppException(ErrorCode.CAPTCHA_FAILED, 'CAPTCHA service is temporarily unavailable');
+    }
+
+    // Reject tokens minted for a different action — prevents cross-action token reuse.
+    if (data.action && data.action !== 'login') {
+      throw new AppException(ErrorCode.CAPTCHA_FAILED, 'CAPTCHA verification failed');
+    }
+
+    const rawScore = parseFloat(this.configService.get<string>('CAPTCHA_MIN_SCORE') ?? '0.5');
```
[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply backend/src/common/captcha/captcha.service.ts:115]: Implemented and verified: reCAPTCHA action is now fail-closed for login (ction must be login), with dedicated unit coverage in captcha.service.spec.ts.

File: frontend/src/hooks/useRecaptchaV3.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+export const useRecaptchaV3 = () => {
+  useEffect(() => {
+    if (!CAPTCHA_ENABLED || CAPTCHA_PROVIDER !== 'google') return;
+    if (document.getElementById('recaptcha-script')) return;
+    const script = document.createElement('script');
+    script.id = 'recaptcha-script';
+    script.src = `https://www.google.com/recaptcha/api.js?render=${CAPTCHA_SITE_KEY}`;
+    script.async = true;
```
[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply frontend/src/hooks/useRecaptchaV3.ts:22]: Addressed from the login UX side: failure states now surface actionable messaging, and the captcha flow is covered by updated frontend wiring plus backend verification tests.

File: backend/src/common/captcha/captcha.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply backend/src/common/captcha/captcha.service.ts:115]: Addressed in latest commit: CAPTCHA score handling is now hardened alongside new captcha service tests. We kept the safe fallback behavior and validated the critical branches with unit tests.

[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply backend/src/common/captcha/captcha.service.ts:115]: Clarification: we now enforce fail-closed action validation (ction must equal login) and added unit coverage in ackend/src/common/captcha/captcha.service.spec.ts.

**Comment by nguyenhoangtin1404** [comment]
Addressed latest CI and review feedback in commit 9c501ff:\n- Added captcha service unit tests and updated auth controller spec.\n- Enforced fail-closed Google action validation for login.\n- Kept timeout and HTTP status handling in CAPTCHA verification paths.\n- Squashed PR history to satisfy commitlint and updated PR title to semantic format.\n\nCurrent required checks have been retriggered by the push.

File: backend/src/common/captcha/captcha.service.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply backend/src/common/captcha/captcha.service.ts:115]: Also improved in 0b79fab: action validation remains fail-closed and is now backed by expanded tests (including score clamp and failure mapping paths).

[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply backend/src/common/captcha/captcha.service.ts:115]: Addressed in 0b79fab: CAPTCHA_MIN_SCORE is now clamped to [0,1] after parsing, and helper tests were extended accordingly.

File: backend/src/auth/auth.controller.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply backend/src/auth/auth.controller.ts]: Updated in 0b79fab: captcha verify now receives the same extractClientIp(req) value used by login audit, so 
emoteip and audit IP stay aligned.

File: frontend/src/hooks/useRecaptchaV3.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nguyenhoangtin1404) [reply frontend/src/hooks/useRecaptchaV3.ts:22]: Implemented in 0b79fab: reCAPTCHA script loading now sets an explicit load-error state and the login page shows a reload action with submit disabled until recovery.

**Comment by nguyenhoangtin1404** [comment]
Added follow-up polish in commit 0b79fab based on latest review threads:\n- unified client IP source for login audit + CAPTCHA remoteip (extractClientIp)\n- mapped CAPTCHA_FAILED to alidation_error in login audit helper\n- clamped CAPTCHA_MIN_SCORE to [0,1]\n- added reCAPTCHA script load-error handling in frontend login UX\n\nValidation rerun locally:\n- backend tests (auth/audit/captcha): pass\n- frontend typecheck: pass
